File: src/compiler/typescript_to_rust/generators/canister_methods/post_upgrade.ts
import {
    JavaScript,
    Rust
} from '../../../../types';
import * as tsc from 'typescript';
import { getCanisterMethodFunctionDeclarationsFromSourceFiles } from '../../../typescript_to_candid/ast_utilities/canister_methods';
import { getUserDefinedInitFunctionParams } from './init';
import { generateIcObject } from '../ic_object';
import { getStableStorageVariableInfos } from './pre_upgrade';
import { getFunctionName } from '../../../typescript_to_candid/ast_utilities/miscellaneous';

export function generateCanisterMethodPostUpgrade(
    sourceFiles: readonly tsc.SourceFile[],
    js: JavaScript,
): Rust {
    const stableStorageVariableInfos = getStableStorageVariableInfos(sourceFiles);

    const icObject: Rust = generateIcObject(stableStorageVariableInfos);

    const initFunctionDeclarations = getCanisterMethodFunctionDeclarationsFromSourceFiles(
        sourceFiles,
        ['Init']
    );

    if (initFunctionDeclarations.length > 1) {
        throw new Error(`Only one init function can be defined`);
    }

    const initFunctionDeclaration: tsc.FunctionDeclaration | undefined = initFunctionDeclarations[0];

    const userDefinedInitFunctionParams = [
        { paramName: 'boa_context', paramType: '&mut boa_engine::Context' },
        ...getUserDefinedInitFunctionParams(initFunctionDeclaration)
    ];

    const postUpgradeFunctionDeclarations = getCanisterMethodFunctionDeclarationsFromSourceFiles(
        sourceFiles,
        ['PostUpgrade']
    );

    if (postUpgradeFunctionDeclarations.length > 1) {
        throw new Error(`Only one PostUpgrade function can be defined`);
    }

    const postUpgradeFunctionDeclaration: tsc.FunctionDeclaration | undefined = postUpgradeFunctionDeclarations[0];

    const developerDefinedPostUpgradeFunctionName = getDeveloperDefinedPostUpgradeFunctionName(postUpgradeFunctionDeclaration);

    return `
        #[ic_cdk_macros::post_upgrade]
        fn post_upgrade(${userDefinedInitFunctionParams.filter((param) => param.paramName !== 'boa_context').map((param) => `${param.paramName}: ${param.paramType}`).join(', ')}) {
            unsafe {
                BOA_CONTEXT_OPTION = Some(boa_engine::Context::default());
                let mut boa_context = BOA_CONTEXT_OPTION.as_mut().unwrap();
        
                ${icObject}
        
                boa_context.register_global_property(
                    "ic",
                    ic,
                    boa_engine::property::Attribute::all()
                );
        
                boa_context.eval(format!(
                    "let exports = {{}}; {compiled_js}",
                    compiled_js = r#"${js}"#
                )).unwrap();

                ${postUpgradeFunctionDeclaration === undefined ? '' : `
                    let exports_js_value = boa_context.eval("exports").unwrap();
                    let exports_js_object = exports_js_value.as_object().unwrap();
        
                    let ${developerDefinedPostUpgradeFunctionName}_js_value = exports_js_object.get("${developerDefinedPostUpgradeFunctionName}", boa_context).unwrap();
                    let ${developerDefinedPostUpgradeFunctionName}_js_object = ${developerDefinedPostUpgradeFunctionName}_js_value.as_object().unwrap();
                    
                    let return_value = ${developerDefinedPostUpgradeFunctionName}_js_object.call(
                        &boa_engine::JsValue::Null,
                        &[
                            ${userDefinedInitFunctionParams.filter((param) => param.paramName !== 'boa_context').map((param) => {
                                return `${param.paramName}.azle_into_js_value(boa_context)`;
                            }).join(',')}
                        ],
                        boa_context
                    ).unwrap();
                `}
            }
        }
    `;
}

function getDeveloperDefinedPostUpgradeFunctionName(initFunctionDeclaration: tsc.FunctionDeclaration | undefined): string {
    if (initFunctionDeclaration === undefined) {
        return '';
    }

    return getFunctionName(initFunctionDeclaration);
}

File: src/compiler/typescript_to_rust/azle_generate/src/generators/canister_methods/system/post_upgrade.rs
use quote::quote;
use swc_ecma_ast::{FnDecl, Program};

use crate::{
    cdk_act::{generators::ic_object, CanisterMethodType, ToTokenStream},
    generators::canister_methods::{functions, method_body},
    ts_ast,
};

pub fn generate_canister_method_system_post_upgrade(
    programs: &Vec<Program>,
) -> proc_macro2::TokenStream {
    let ic_object = ic_object::generate_ic_object(programs);

    let post_upgrade_fn_decls = ts_ast::program::get_canister_method_type_fn_decls(
        programs,
        &CanisterMethodType::PostUpgrade,
    );

    if post_upgrade_fn_decls.len() > 1 {
        panic!("Only one PostUpgrade function can be defined");
    }

    let post_upgrade_fn_decl_option = post_upgrade_fn_decls.get(0);

    let post_upgrade_params = generate_post_upgrade_params(&post_upgrade_fn_decl_option);

    let call_to_post_upgrade_js_function =
        generate_call_to_post_upgrade_js_function(&post_upgrade_fn_decl_option);

    quote! {
        #[ic_cdk_macros::post_upgrade]
        fn _azle_post_upgrade(#(#post_upgrade_params),*) {
            unsafe {
                BOA_CONTEXT_OPTION = Some(boa_engine::Context::default());
                let mut _azle_boa_context = BOA_CONTEXT_OPTION.as_mut().unwrap();

                _azle_boa_context.eval(format!(
                    "let exports = {{}}; {compiled_js}",
                    compiled_js = STABLE_STORAGE_JS
                )).unwrap();

                let _azle_stable_storage: (String,) = ic_cdk::storage::stable_restore().unwrap();
                let _azle_stable_storage_json_string = _azle_stable_storage.0;

                let _azle_stable_storage_deserialize_exports_js_value = _azle_boa_context.eval("exports").unwrap();
                let _azle_stable_storage_deserialize_exports_js_object = _azle_stable_storage_deserialize_exports_js_value.as_object().unwrap();

                let _azle_stable_storage_deserialize_function_js_value = _azle_stable_storage_deserialize_exports_js_object.get("stable_storage_deserialize", &mut _azle_boa_context).unwrap();
                let _azle_stable_storage_deserialize_function_js_object = _azle_stable_storage_deserialize_function_js_value.as_object().unwrap();

                let _azle_stable_storage = _azle_stable_storage_deserialize_function_js_object.call(
                    &boa_engine::JsValue::Null,
                    &[
                        _azle_stable_storage_json_string.azle_into_js_value(&mut _azle_boa_context)
                    ],
                    &mut _azle_boa_context
                ).unwrap();

                #ic_object

                _azle_boa_context.register_global_property(
                    "ic",
                    ic,
                    boa_engine::property::Attribute::all()
                );

                _azle_boa_context.eval(format!(
                    "{compiled_js}",
                    compiled_js = MAIN_JS
                )).unwrap();

                #call_to_post_upgrade_js_function
            }
        }
    }
}

fn generate_post_upgrade_params(
    post_upgrade_fn_decl_option: &Option<&FnDecl>,
) -> Vec<proc_macro2::TokenStream> {
    if let Some(post_upgrade_fn_decl) = post_upgrade_fn_decl_option {
        // TODO that params data structures can have the name, the type, and both strings and idents as necessary
        functions::build_params(&post_upgrade_fn_decl)
            .iter()
            .map(|param| param.to_token_stream())
            .collect()
    } else {
        vec![]
    }
}

fn generate_call_to_post_upgrade_js_function(
    post_upgrade_fn_decl_option: &Option<&FnDecl>,
) -> proc_macro2::TokenStream {
    if let Some(post_upgrade_fn_decl) = post_upgrade_fn_decl_option {
        method_body::generate_call_to_js_function(post_upgrade_fn_decl)
    } else {
        quote!()
    }
}
